File: scripts/algorithms/M/Minimum Swaps to Arrange a Binary Grid/Minimum Swaps to Arrange a Binary Grid.java
class Solution {
    public int minSwaps(int[][] grid) {
        int n = grid.length, ans = 0, cur = 0;
        for (int k = 0; k < n - 1; k++){ // looking for the fitting row for row k
            for (int i = k; i < n; i++){ // start from row k looking downward
                for (int j = k + 1; j < n; j++){ // all cell after and at k + 1 must be 0
                    if (grid[i][j] == 1)
                        break;
                    if (j < n - 1)
                        continue;
                    for (int m = i; m > k; m--){ // j == n - 1 here, so we found a valid row
                        int[] tmp = grid[m - 1];  // swap it into the correct row - row k
                        grid[m - 1] = grid[m];
                        grid[m] = tmp;
                        ans++;
                    }
                    i = n;
                }
                if (i == n - 1) // i reaches the end and did not find a fitting row, return -1
                    return -1;
            }
        }
        return ans;
    }
}

File: scripts/algorithms/M/Minimum Swaps to Arrange a Binary Grid/Minimum Swaps to Arrange a Binary Grid.py
# Runtime: 1405 ms (Top 5.77%) | Memory: 15.2 MB (Top 23.08%)
class Solution:
    def minSwaps(self, grid) -> int:
        n = len(grid)
        max_right = [-1] * n
        for r, row in enumerate(grid):
            for c in range(n - 1, -1, -1):
                if row[c] == 1:
                    max_right[r] = c
                    break
        if all(v <= i for i, v in enumerate(sorted(max_right))):
            swaps = 0
            i = 0
            while i < n:
                while i < n and max_right[i] <= i:
                    i += 1
                if i == n:
                    break
                j = i
                while j < n and max_right[j] > i:
                    j += 1
                swaps += j - i
                max_right[i], max_right[i + 1: j + 1] = (max_right[j],
                                                         max_right[i: j])
                i += 1
            return swaps
        return -1